[RFC] Proposer-evaluted KV

tbg opened the pull request:
[Read the latest version here](https://github.com/cockroachdb/cockroach/blob/tschottdorf/leader-evaluated-raft/docs/RFCS/leader_evaluated_raft.md).

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="35" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/6166)

<!-- Reviewable:end -->

bdarnell commented:
Review status: 0 of 1 files reviewed at latest revision, 10 unresolved discussions.

---

*[docs/RFCS/leader_evaluated_raft.md, line 95 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFlgN2CU7OLU5qyfhVh:-KFlgN2DFnBLFBlJhbSi:305738429) ([raw file](https://github.com/cockroachdb/cockroach/blob/fca7031b74043afdf2d8ccb45649dbae36c8630c/docs/RFCS/leader_evaluated_raft.md#L95)):*
This still seems like a lot of logic to have post-raft. Would the recent lease changes have been safe in this model, or would we have needed an approach more like the one in #5985 for that one? 

My initial reaction on seeing this list was that this is exactly the stuff that we may need to change and we haven't gained much if it's just a way to make changes to the relatively simple commands that aren't on this list. But after looking at the list of migrations considered in #5985 it does seem to cover most of them. And even for these commands the bulk of the work could still be done with the mvcc batch. For instance, only the call to `Store.SplitRange` needs to be covered here; the rest of the `splitTrigger` can be run pre-raft. And the portions of these commands that remain post-raft may mainly operate on things that are outside the raft "state machine" so compatibility requirements may not be as strict. 

---

*[docs/RFCS/leader_evaluated_raft.md, line 96 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFlooZLaQQOAad-El9D:-KFlooZMWat9Yuk33VDC:-1412174334) ([raw file](https://github.com/cockroachdb/cockroach/blob/fca7031b74043afdf2d8ccb45649dbae36c8630c/docs/RFCS/leader_evaluated_raft.md#L96)):*
What would be done here for the abort cache? Isn't it just KV pairs with no in-memory state or side effects?

---

*[docs/RFCS/leader_evaluated_raft.md, line 178 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFlcc1nub0RcPfpjjfm:-KFlcc1ohX1BD1jnz-l-:-1551554599) ([raw file](https://github.com/cockroachdb/cockroach/blob/fca7031b74043afdf2d8ccb45649dbae36c8630c/docs/RFCS/leader_evaluated_raft.md#L178)):*
We had a pure go batch implementation prior to 7d4e2fced. Would it be better to revive that than to wrap a batch with something that saves a second copy of all writes?

---

*[docs/RFCS/leader_evaluated_raft.md, line 208 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFldLC4c_AAKZVsaXwO:-KFldLC5jmXntOqeNT6g:2088088137) ([raw file](https://github.com/cockroachdb/cockroach/blob/fca7031b74043afdf2d8ccb45649dbae36c8630c/docs/RFCS/leader_evaluated_raft.md#L208)):*
Some more detail: The `RaftCmd` would also include a term and log index, to be populated with the leader's latest log information when it is proposed. The writes would only be applied if it committed at that index; they would be dropped (and reproposed) if they committed at any other index. This would only be a possibility when the lease holder is not the raft leader (and even then it's unlikely in stable lease situations). We may be able to be even stricter about colocating the two roles (in which case we could perhaps remove raft-level replays completely), but I'm wary of forcing too many raft elections.

---

*[docs/RFCS/leader_evaluated_raft.md, line 232 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFliuLLzFae0-bREtPV:-KFliuLMEj3iQi6K_i_3:-2136252056) ([raw file](https://github.com/cockroachdb/cockroach/blob/fca7031b74043afdf2d8ccb45649dbae36c8630c/docs/RFCS/leader_evaluated_raft.md#L232)):*
Duplicate heading

---

*[docs/RFCS/leader_evaluated_raft.md, line 238 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFlkBRUv5ozxqqjjjkS:-KFlkBRUv5ozxqqjjjkT:626929332) ([raw file](https://github.com/cockroachdb/cockroach/blob/fca7031b74043afdf2d8ccb45649dbae36c8630c/docs/RFCS/leader_evaluated_raft.md#L238)):*
Downreplicating to a single copy scares me. Partially for the obvious concern that an ill-timed disk failure could lead to data loss, but mainly because that operation is not well tested and is likely to encounter edge cases around quorums. A single-replica configuration in a multi-node cluster may make the rebalancer do interesting things (e.g. I'm not sure if you can stop it from creating a temporary second replica as it moves things around).

---

*[docs/RFCS/leader_evaluated_raft.md, line 251 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFljlJKOANVPQX5bANC:-KFljlJLTCBhvROyMfql:1016882769) ([raw file](https://github.com/cockroachdb/cockroach/blob/fca7031b74043afdf2d8ccb45649dbae36c8630c/docs/RFCS/leader_evaluated_raft.md#L251)):*
We'll eventually need something like the scheme in #5985 for migrations of what remains post-raft. So anything done in that direction, while complex, would not be wasted work. In particular, the Freeze/Unfreeze pair of RPCs I proposed would make it possible to make this switch without downreplicating to a single copy.

---

*[docs/RFCS/leader_evaluated_raft.md, line 254 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFln3HlHikvk-iS2zSi:-KFln3HmhOu2vd_dMnJJ:1684763855) ([raw file](https://github.com/cockroachdb/cockroach/blob/fca7031b74043afdf2d8ccb45649dbae36c8630c/docs/RFCS/leader_evaluated_raft.md#L254)):*
More drawbacks:
* While the amount of code downstream of raft is drastically reduced, it is not zero, so we will still need something like #5985 eventually. But hopefully this will not be something we need to use often.
* "Dumb" raft commands generally need to be strictly serialized by the command queue, while "smart" ones can be pipelined more effectively. We currently enforce serialization anyway, but this change removes the option of pipelining. 

---

*[docs/RFCS/leader_evaluated_raft.md, line 256 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFliSp9p1bO8XcLQlyD:-KFliSp9p1bO8XcLQlyE:335750459) ([raw file](https://github.com/cockroachdb/cockroach/blob/fca7031b74043afdf2d8ccb45649dbae36c8630c/docs/RFCS/leader_evaluated_raft.md#L256)):*
`DeleteRange` is an extreme case, but I would guess that most commands will have some inflation here (would be good to measure with a quick prototype). We're saving the CPU work and disk reads by not repeating those operations on each replica, but we'll probably be increasing network traffic and disk writes. 

Splits will also be large if the abort cache is copied into KV Write pairs.

---

*[docs/RFCS/leader_evaluated_raft.md, line 260 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFlo46NB0WhKo_kzeTq:-KFlo46NB0WhKo_kzeTr:1686778091) ([raw file](https://github.com/cockroachdb/cockroach/blob/fca7031b74043afdf2d8ccb45649dbae36c8630c/docs/RFCS/leader_evaluated_raft.md#L260)):*
s/and no//

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166)*
<!-- Sent from Reviewable.io -->

tbg committed:
Address first round of comments

Misc edits and flesh out

- remaining post-Raft code
- replay protection
- migration strategy
- drawbacks
- unresolved questions

tschottdorf commented:
Added a second commit to address @bdarnell's initial round of comments.

---

Review status: 0 of 1 files reviewed at latest revision, 10 unresolved discussions, some commit checks pending.

---

*[docs/RFCS/leader_evaluated_raft.md, line 95 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFlgN2CU7OLU5qyfhVh:-KFmJ3mXCTf54v6dbMmk:-1821691220) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1133414ccfca7678b600a2f712c31dd56cebf18/docs/RFCS/leader_evaluated_raft.md#L95)):*
> And the portions of these commands that remain post-raft may mainly operate on things that are outside the raft "state machine" so compatibility requirements may not be as strict.

Yes, exactly. `LeaderLease` is in this list only because we have in-memory state on `*Replica` which we need to change (and there's this bit about gossipping some stuff if the new leader changes and updating the timestamp cache, etc). It's not the intention that all of "(*Replica).LeaderLease` move, though we will want the `Lease` proto in this trigger, so its migration would be a little harder potentially.

I added more detail on the individual triggers in the mock proto below.

---

*[docs/RFCS/leader_evaluated_raft.md, line 96 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFlooZLaQQOAad-El9D:-KFmR34CXo6JP3lw-2dp:16232753) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1133414ccfca7678b600a2f712c31dd56cebf18/docs/RFCS/leader_evaluated_raft.md#L96)):*
You're right, nothing to do there. Removed.

---

*[docs/RFCS/leader_evaluated_raft.md, line 178 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFlcc1nub0RcPfpjjfm:-KFmULW_zJJ4Qi44SA7E:-1454511958) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1133414ccfca7678b600a2f712c31dd56cebf18/docs/RFCS/leader_evaluated_raft.md#L178)):*
Added a discussion of this to the `Unresolved questions` section. The best implementation would be RocksDB exposing a read-only (zero-copy) view of queued mutations; everything else we'd have to benchmark.

---

*[docs/RFCS/leader_evaluated_raft.md, line 208 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFldLC4c_AAKZVsaXwO:-KFmUzJpel4T3V77fYN3:742293137) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1133414ccfca7678b600a2f712c31dd56cebf18/docs/RFCS/leader_evaluated_raft.md#L208)):*
Inserted.

---

*[docs/RFCS/leader_evaluated_raft.md, line 232 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFliuLLzFae0-bREtPV:-KFmV4926adVymIVSABS:761715002) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1133414ccfca7678b600a2f712c31dd56cebf18/docs/RFCS/leader_evaluated_raft.md#L232)):*
did `s/Implement/Migr/`

---

*[docs/RFCS/leader_evaluated_raft.md, line 251 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFljlJKOANVPQX5bANC:-KFmVXbgS73mxC-C1dWy:378869330) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1133414ccfca7678b600a2f712c31dd56cebf18/docs/RFCS/leader_evaluated_raft.md#L251)):*
Ah, `{F,Unf}reeze` does make a lot of sense here; it had dropped off my radar. Updated this section.

---

*[docs/RFCS/leader_evaluated_raft.md, line 254 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFln3HlHikvk-iS2zSi:-KFmdkt-ely1KkUOgL7d:-206782473) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1133414ccfca7678b600a2f712c31dd56cebf18/docs/RFCS/leader_evaluated_raft.md#L254)):*
Added. Can you elaborate on not being able to pipeline (ideally with a simple example)? I wasn't quite clear on the discussion in #5985 - it seemed to me that it would be as difficult then as it is now, so I'm likely missing the big picture.
The idea would be that we would want to shove a bunch of overlapping writes into Raft simultaneously, correct? Since we can avoid Raft reordering now (or at least with changes proposed in this RFC), the problem boils down to having all intermediate states which would result from applying the prefixes of the existing overlapping writes (including side effects) available to execute on top of. This RFC appears to operate completely below that layer.

---

*[docs/RFCS/leader_evaluated_raft.md, line 256 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFliSp9p1bO8XcLQlyD:-KFmYIFy9H_2S_NalEgW:967249843) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1133414ccfca7678b600a2f712c31dd56cebf18/docs/RFCS/leader_evaluated_raft.md#L256)):*
Added. The abortCache is (or at least should be) pretty small these days.

---

*[docs/RFCS/leader_evaluated_raft.md, line 260 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFlo46NB0WhKo_kzeTq:-KFmavlC1PqVxxammbJy:120223049) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1133414ccfca7678b600a2f712c31dd56cebf18/docs/RFCS/leader_evaluated_raft.md#L260)):*
Done. (-ish)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-:-KFmexk1l2Prx6WOfELe:-2015535431)*
<!-- Sent from Reviewable.io -->

tbg committed:
[RFC] Leader-evaluted Raft

This is a hastily written, initial draft released early to get early feedback
for more refined iterations.

[ci skip]

petermattis commented:
Review status: 0 of 1 files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/leader_evaluated_raft.md, line 178 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFlcc1nub0RcPfpjjfm:-KFnaqDV3qZHeVTPSoer:-1973219231) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1133414ccfca7678b600a2f712c31dd56cebf18/docs/RFCS/leader_evaluated_raft.md#L178)):*
There is a significant stability and performance risk in moving away from the RocksDB WriteBatch as we would have to rewrite the "merged" iterator used for batches and the existing WriteBatch code has been tuned while a new implementation would not.

Underneath the hood, a RocksDB WriteBatch stores all of the writes in a C++ `string` (or the equivalent). This string is persisted to the RocksDB WAL when the WriteBatch is written. And note that all mutating operations in RocksDB are internally converted into WriteBatches. This means that the format of the internal WriteBatch buffer is stable. 

I see two reasonable options here. One is that we simply pass around the internal WriteBatch buffer. RocksDB already has an interface for retrieving this internal data and adding an interface to create a WriteBatch from such a buffer would be straightforward.

The second option would be to iterate over the WriteBatch mutations and build up our own set of mutations to pass over the wire. This would be slightly less efficient, but would divorce us from RocksDB.

My preference is for the first option (using the internal `WriteBatch::Data()`). We can note the format of this data in the RaftCmd to allow for future versioning (e.g. if we ever move away from RocksDB). The format of this internal data is reasonably simple to parse if we ever need to do so.

---

*[docs/RFCS/leader_evaluated_raft.md, line 361 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFnd7jm4OQhwDAaPlrI:-KFnd7jn5a8TLGZH0GKg:2023257134) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff7bb96c8d04ff03e1ee2d7714997b23b99083be/docs/RFCS/leader_evaluated_raft.md#L361)):*
There is a significant stability and performance risk in moving away from the RocksDB WriteBatch as we would have to rewrite the "merged" iterator used for batches and the existing WriteBatch code has been tuned while a new implementation would not.

Underneath the hood, a RocksDB WriteBatch stores all of the writes in a C++ `string`. This string is persisted to the RocksDB WAL when the WriteBatch is written. Note that all mutating operations in RocksDB are internally converted into WriteBatches (i.e. this code path is heavily optimized and the format is stable).

I see two reasonable options here. One is that we simply pass around the internal WriteBatch buffer. RocksDB already has an interface for retrieving this internal data (`WriteBatch::Data()`) and adding an interface to create a WriteBatch from such a buffer would be straightforward.

The second option would be to iterate over the WriteBatch mutations and build up our own set of mutations to pass over the wire. This would be slightly less efficient, but would divorce us from RocksDB.

My preference is for the first option (using the internal `WriteBatch::Data()`). We can note the format of this data in the RaftCmd to allow for future versioning (e.g. if we ever move away from RocksDB). The format of this internal data is reasonably simple to parse if we ever need to do so.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 1 files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/leader_evaluated_raft.md, line 361 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFnd7jm4OQhwDAaPlrI:-KFndeCJBjcNSw4M3Xj0:-1769294161) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff7bb96c8d04ff03e1ee2d7714997b23b99083be/docs/RFCS/leader_evaluated_raft.md#L361)):*
Ah, I see `WriteBatch::WriteBatch(const std::string& rep)` already exists. There are no changes to RocksDB necessary to pass around the internal `WriteBatch` buffer and reconstitute it into a `WriteBatch` on the receiving side.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 1 files reviewed at latest revision, 11 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/leader_evaluated_raft.md, line 178 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFlcc1nub0RcPfpjjfm:-KFnx91O8z-exBg5dokT:-2121932563) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1133414ccfca7678b600a2f712c31dd56cebf18/docs/RFCS/leader_evaluated_raft.md#L178)):*
This comment was accidentally duplicated here. Ignore and follow up on the version below.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/leader_evaluated_raft.md, line 280 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFof8fX8-JGcTM7rEkS:-KFof8fX8-JGcTM7rEkT:-1974263332) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff7bb96c8d04ff03e1ee2d7714997b23b99083be/docs/RFCS/leader_evaluated_raft.md#L280)):*
s/a a/a/

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/leader_evaluated_raft.md, line 254 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFln3HlHikvk-iS2zSi:-KFp32Eeeu7QVVw2Vr22:-1230636195) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1133414ccfca7678b600a2f712c31dd56cebf18/docs/RFCS/leader_evaluated_raft.md#L254)):*
Ah yes, the pipelining problems have to do with raft-level reordering so as long as we can solve that one then we can pipeline just as easily (or not) as we could before. 

---

*[docs/RFCS/leader_evaluated_raft.md, line 361 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFnd7jm4OQhwDAaPlrI:-KFp3_QVZtGc7_MaIvT4:-775429434) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff7bb96c8d04ff03e1ee2d7714997b23b99083be/docs/RFCS/leader_evaluated_raft.md#L361)):*
If we can get the `WriteBatch` data in and out of rocksdb then that sounds ideal. My biggest concern would be about the compatibility guarantees that rocksdb makes (or doesn't) about this format. If we use this then I think we should write and test a parser even if we don't use it in production, so we will notice if rocksdb ever changes their format (if they change then we'll need to convert their new format to the old one for compatibility). 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166)*
<!-- Sent from Reviewable.io -->

tbg committed:
Second round of comments

Flesh out

- use of RocksDB WriteBatch for zero-copy serialization
- concretize implementation strategy
- updated discussion on removal of write-write blocking

[ci skip]

tschottdorf commented:
Updated, PTAL. I feel that this PR has split up relatively well. We have (depending on how you count) 7 well-contained prerequisites which could in theory all be worked on in parallel (the one needing most thought likely being the Freeze/Unfreeze part, which is queued first) and which would (mostly) be useful independently. The remaining "forklift" portion which actually moves the bulk processing pre-Raft is still going to take some effort, but it does appear to me that the bulk of the work is in those prerequisites.

---

Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions.

---

*[docs/RFCS/leader_evaluated_raft.md, line 254 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFln3HlHikvk-iS2zSi:-KFsoGYarUA3qCM9-1uE:1356123053) ([raw file](https://github.com/cockroachdb/cockroach/blob/b1133414ccfca7678b600a2f712c31dd56cebf18/docs/RFCS/leader_evaluated_raft.md#L254)):*
Updated the discussion.

---

*[docs/RFCS/leader_evaluated_raft.md, line 280 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFof8fX8-JGcTM7rEkS:-KFsnI93g0ljo44kRHBz:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff7bb96c8d04ff03e1ee2d7714997b23b99083be/docs/RFCS/leader_evaluated_raft.md#L280)):*
Done.

---

*[docs/RFCS/leader_evaluated_raft.md, line 361 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFnd7jm4OQhwDAaPlrI:-KFsj-_wNuzNkblzH7tI:-1984906616) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff7bb96c8d04ff03e1ee2d7714997b23b99083be/docs/RFCS/leader_evaluated_raft.md#L361)):*
Here's the format (from `c-rocksdb/internal/db/write_batch.cc`; note that this is copied from a comment, so it could already have rotted):

```
// WriteBatch::rep_ :=
//    sequence: fixed64
//    count: fixed32
//    data: record[count]
// record :=
//    kTypeValue varstring varstring
//    kTypeDeletion varstring
//    kTypeSingleDeletion varstring
//    kTypeMerge varstring varstring
//    kTypeColumnFamilyValue varint32 varstring varstring
//    kTypeColumnFamilyDeletion varint32 varstring varstring
//    kTypeColumnFamilySingleDeletion varint32 varstring varstring
//    kTypeColumnFamilyMerge varint32 varstring varstring
// varstring :=
//    len: varint32
//    data: uint8[len]
```

For reference, LevelDB has the following:

```
// WriteBatch::rep_ :=
//    sequence: fixed64
//    count: fixed32
//    data: record[count]
// record :=
//    kTypeValue varstring varstring         |
//    kTypeDeletion varstring
// varstring :=
//    len: varint32
//    data: uint8[len]
```

Doesn't seem unlikely that `RocksDB` will diverge more as they add features, so I agree with writing the parser (also to not make it even harder to switch to another storage engine in the future).

I updated the RFC to reflect that this is the suggested implementation.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-:-KFssRbmHv_KlrHGIiEi:41162774)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions.

---

*[docs/RFCS/leader_evaluated_raft.md, line 361 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFnd7jm4OQhwDAaPlrI:-KFtBqvM4l-QVuhkSzkT:-1072069335) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff7bb96c8d04ff03e1ee2d7714997b23b99083be/docs/RFCS/leader_evaluated_raft.md#L361)):*
While the RocksDB WriteBatch format might certainly be enhanced in the future, any enhancements are likely to be backward compatible as a non-compatible change would require rewriting the entries in the WAL. I think writing our own parser and verification of this format would be good, though it shouldn't be a blocker/prerequisite for the rest of the work in this RFC. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 1 files reviewed at latest revision, 7 unresolved discussions.

---

*[docs/RFCS/leader_evaluated_raft.md, line 219 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFtEHEKn3taPNeuzRZ0:-KFtEHEKn3taPNeuzRZ1:415601275) ([raw file](https://github.com/cockroachdb/cockroach/blob/460f9c23e232d9de61558807d83ebc96c288d4f7/docs/RFCS/leader_evaluated_raft.md#L219)):*
There is another nice effect from using the `WriteBatch`: instead of marshaling and unmarshaling individual keys and values, we'll be passing around a blob. While investigating batch insert performance I added some instrumentation around `RaftCommand` marshaling and noticed:

```
raft command marshal:   20002: 6.2ms
raft command unmarshal: 20002: 29.1ms
raft command marshal:   20002: 6.0ms
raft command unmarshal: 20002: 12.4ms
raft command marshal:   20002: 6.0ms
raft command unmarshal: 20002: 42.6ms
```

The first number (`20002`) is the number of requests in the `BatchRequest` and the second is the time it took to marshal/unmarshal. The marshaled data size is almost exactly 1MB.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166)*
<!-- Sent from Reviewable.io -->

tbg committed:
Add serialization numbers

tschottdorf commented:
Review status: 0 of 1 files reviewed at latest revision, 7 unresolved discussions, some commit checks pending.

---

*[docs/RFCS/leader_evaluated_raft.md, line 361 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFnd7jm4OQhwDAaPlrI:-KFtMpbGCk-vaWe0os_k:-1236676892) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff7bb96c8d04ff03e1ee2d7714997b23b99083be/docs/RFCS/leader_evaluated_raft.md#L361)):*
I added that to the implementation strategy.

---

*[docs/RFCS/leader_evaluated_raft.md, line 219 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFtEHEKn3taPNeuzRZ0:-KFtK7MJOIyO9xTGDhYS:-310247768) ([raw file](https://github.com/cockroachdb/cockroach/blob/460f9c23e232d9de61558807d83ebc96c288d4f7/docs/RFCS/leader_evaluated_raft.md#L219)):*
Yes, check out the current revision which points this out as one of the advantages. Good to have these numbers - Wove them into the draft.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/leader_evaluated_raft.md, line 361 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFnd7jm4OQhwDAaPlrI:-KFtcU0hBzAgZGPiUCcj:-2065834957) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff7bb96c8d04ff03e1ee2d7714997b23b99083be/docs/RFCS/leader_evaluated_raft.md#L361)):*
We can trust rocksdb to be backwards-compatible in any changes they make, but we require *bidirectional* compatibility. We create a batch on the leader and ship it off to the follower to be applied, and the follower might be running either an older or a newer version of the code. If they add a new record type, we either need to be able to guarantee that the new record types won't be used in our case, or translate them back to something the old version can understand. 

---

*[docs/RFCS/leader_evaluated_raft.md, line 219 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFtEHEKn3taPNeuzRZ0:-KFtdGj5K3wXNavLuvb8:-762343511) ([raw file](https://github.com/cockroachdb/cockroach/blob/460f9c23e232d9de61558807d83ebc96c288d4f7/docs/RFCS/leader_evaluated_raft.md#L219)):*
Unmarshaling is much slower than marshaling because it copies each `bytes` field into a separately allocated array, while marshaling computes the total size needed and does one big allocation. It might be worth exploring a non-copying option for gogoproto that simply sets the bytes field to a slice from the input data. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-:-KFte133UD2m7XlhFKOd:1417088721)*
<!-- Sent from Reviewable.io -->

tbg committed:
Add comment by @bdarnell on WriteBatch compatibility

tschottdorf commented:
Seems that there's a good consensus so far - @spencerkimball and anyone else interested, I'll keep this open until Monday night ET (and longer if there's discussion). PTAL.

Once merged, I'll create a roadmap (issues for each points outlined in the implementation strategy below and the remainder (and perhaps a label)), but we should also determine how long we're hoping to get this done and who will work on it.

---

Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions, some commit checks pending.

---

*[docs/RFCS/leader_evaluated_raft.md, line 361 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFnd7jm4OQhwDAaPlrI:-KFunfECZDirOaBxQs67:1610000531) ([raw file](https://github.com/cockroachdb/cockroach/blob/ff7bb96c8d04ff03e1ee2d7714997b23b99083be/docs/RFCS/leader_evaluated_raft.md#L361)):*
Added.

---

*[docs/RFCS/leader_evaluated_raft.md, line 219 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFtEHEKn3taPNeuzRZ0:-KFuoFv1gwXJldwUbVn7:-238287731) ([raw file](https://github.com/cockroachdb/cockroach/blob/460f9c23e232d9de61558807d83ebc96c288d4f7/docs/RFCS/leader_evaluated_raft.md#L219)):*
Good idea, but doesn't fit here - is there any issue to add this to, or could you file it otherwise?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-:-KFuoNUBOI-L63PvupuS:190179589)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 0 of 1 files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[docs/RFCS/leader_evaluated_raft.md, line 219 \[r7\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFtEHEKn3taPNeuzRZ0:-KFurYMJJBMBBOjkG-O3:-531379582) ([raw file](https://github.com/cockroachdb/cockroach/blob/460f9c23e232d9de61558807d83ebc96c288d4f7/docs/RFCS/leader_evaluated_raft.md#L219)):*
#6208

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166)*
<!-- Sent from Reviewable.io -->

JackKrupansky commented:
Is the presumption here that an upgrade/migration of a node must be performed while a transaction is in progress?

I mean, why not use the following model:

1. Mark the node as frozen. No new transactions involving ranges replicated to the node can begin.
2. Let current transactions run to completion.
3. If step 2 takes too long, have the option of aborting any remaining transactions. No new transactions would be started since the node was frozen.
4. Upgrade/migrate the node, restart.
5. Unfreeze the node.

BTW, with Cassandra, if the application is reading and writing with quorum consistency, any single node could be brought down without breaking quorum, assuring absolute non-stop operation. The node would then catch up due to hinted hand-off, provided that the down time was small, under an hour or whatever the handoff timeout grace period is.

It would be nice if there was an application return code indicating that the server is in maintenance mode and that the application should hold off new transactions for some interval of time, rather than the application simply getting some error and mindlessly retrying the operation or assuming that the server is in a bad state rather than simply being maintained.

tschottdorf commented:
No, this happens below the level of transactions. The primary source of complication is that we keep byte-wise consistent replicas (a problem Cassandra does not have) and that the post-raft evaluation of commands makes the bulk of code live downstream of the point at which byte-wise output must be identical.
The goal for (most) migrations is a rolling cluster restart (which with appropriately configured clients) is completely interruption free.

---

Review status: 0 of 1 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-:-KFwXTQbe_1cEIgQx4JK:234421088)*
<!-- Sent from Reviewable.io -->

spencerkimball commented on docs/RFCS/leader_evaluated_raft.md:
```diff
@@ -0,0 +1,425 @@
+
+* any key-value writes: For example the outcome of a `Put`, `DeleteRange`, `CPut`, `BeginTransaction`, or `LeaderLease`.
+  Some alternatives exist: writes could be encoded as `[]MVCCKeyValue`, or even
+  as post-MVCC raw (i.e. opaque) key-value pairs. We'll assume the former for
```
Why assume []MVCCKeyValue for the "effects" contents? Then we're tying up potential for migration in how we encode MVCC information. I think it's more prudent to send "post-MVCC raw" key-value pairs.

spencerkimball commented on docs/RFCS/leader_evaluated_raft.md:
```diff
@@ -0,0 +1,425 @@
+  repeated oneof effect {
+    # Applied in field and slice order.
+    optional Writes writes = ...; # see MVCC section for Writes type
+    # Carries out the following:
```
In the proposed RaftCmd, I assume you're just going to use the various Trigger protobufs we already have? Let's not create new types for this struct.

spencerkimball commented on docs/RFCS/leader_evaluated_raft.md:
```diff
@@ -0,0 +1,425 @@
+
+With the new code, we want mostly the same, but with two important changes:
+
+* we do not actually intent to commit the batch (though we could do this on the
```
s/intent/intend/

spencerkimball commented on docs/RFCS/leader_evaluated_raft.md:
```diff
@@ -0,0 +1,425 @@
+
+* we do not actually intent to commit the batch (though we could do this on the
+  proposing node to avoid overhead)
+* we need to construct the `Writes` field, which (in some way or
```
I don't think you'll need any of the changes proposed in this MVCC section (as you point out in the section proposing the use of RocksDB's `WriteBatch` format).

spencerkimball commented on docs/RFCS/leader_evaluated_raft.md:
```diff
@@ -0,0 +1,425 @@
+
+## Raft command application
+
+This will be a new, straightforward facility which applies a `RaftCmd` against
```
I propose that we send the serialized bytes from the `WriteBatch` as a `bytes` parameter in `RaftCmd` and then pass that directly to a new cgo shim for our engine which would deserialize as necessary down in C++ land. In C++, we'd fork the current RocksDB `WriteBatch` protobuf and use that in our C++ `Batch` implementation. As long as we're using RocksDB, it'll be a trivial assignment of bytes. Other engines would require a translation. But by forking the RocksDB `WriteBatch`, we'll stabilize our own version as well as keep all of these details opaque to the Go implementation.

petermattis replied:
I'm not sure what you mean by "fork the current RocksDB `WriteBatch`". FYI, `WriteBatch` is a C++ class, not a protobuf. And we don't require any changes to it for this proposal as it already exposes the encoded operations as a `string` and allows construction of a `WriteBatch` given a `string` of encoded operations.

spencerkimball commented on docs/RFCS/leader_evaluated_raft.md:
```diff
@@ -0,0 +1,425 @@
+
+## Replay protection
+
+Replay protection is naively an issue because applying a logicless batch of
```
Let's plan an audit of the code to remove anything which I added in the big transaction semantics refactoring to notice Raft replays. I just want to make sure we don't have code to identify Raft replays living in there once we move this stuff in front of Raft.

tschottdorf commented:
Any follow-up discussions are best held in the appropriate subissue: https://github.com/cockroachdb/cockroach/labels/leader-proposed-raft

---

Review status: 0 of 1 files reviewed at latest revision, 9 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/leader_evaluated_raft.md, line 71 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFkX4GAbVSXL3RNV8H4-r17-71:-KGNfajgW0F66Ty0T0Lb:-1787687110) ([raw file](https://github.com/cockroachdb/cockroach/blob/625193eee94b4fd02310b88ccea90aa121a085e5/docs/RFCS/leader_evaluated_raft.md#L71)):*
The above is an option. The later discussion settles on `WriteBatch` as elaborated on later (which is a post-MVCC raw option).

---

*[docs/RFCS/leader_evaluated_raft.md, line 98 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFkX4GAbVSXL3RNV8H4-r17-98:-KGNfmZA8u7088JYJJbo:2123900799) ([raw file](https://github.com/cockroachdb/cockroach/blob/625193eee94b4fd02310b88ccea90aa121a085e5/docs/RFCS/leader_evaluated_raft.md#L98)):*
We have way more side effects and the triggers that we have make nontrivial use of data which may no more be available (for example, you won't know the txn from the WriteBatch). I don't think we can really see what we need a priori, got to do the refactor first. I'm sure we'll see new protos here.

---

*[docs/RFCS/leader_evaluated_raft.md, line 175 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFkX4GAbVSXL3RNV8H4-r17-175:-KGNg7fsVH-sPxnBMAy5:185001303) ([raw file](https://github.com/cockroachdb/cockroach/blob/625193eee94b4fd02310b88ccea90aa121a085e5/docs/RFCS/leader_evaluated_raft.md#L175)):*
Note to self: fix this typo when adjusting the status of this RFC.

---

*[docs/RFCS/leader_evaluated_raft.md, line 177 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFkX4GAbVSXL3RNV8H4-r17-177:-KGNgHqqwRbtR_3ULbx-:1727370400) ([raw file](https://github.com/cockroachdb/cockroach/blob/625193eee94b4fd02310b88ccea90aa121a085e5/docs/RFCS/leader_evaluated_raft.md#L177)):*
Good point, made a comment on the issue.

---

*[docs/RFCS/leader_evaluated_raft.md, line 274 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KFkX4GAbVSXL3RNV8H4-r17-274:-KGR4f_Lv2gyFuy3NAl3:1717857520) ([raw file](https://github.com/cockroachdb/cockroach/blob/625193eee94b4fd02310b88ccea90aa121a085e5/docs/RFCS/leader_evaluated_raft.md#L274)):*
Added to #6287.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-:-KGR4oUl5G_QgTkkwk2M:-1048867362)*
<!-- Sent from Reviewable.io -->

vivekmenezes commented:
Review status: 0 of 1 files reviewed at latest revision, 10 unresolved discussions, all commit checks successful.

---

*[docs/RFCS/leader_evaluated_raft.md, line 35 \[r17\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166#-KGSFBgC0ZNHZxaXrWDg:-KGSFBgC0ZNHZxaXrWDh:1316490142) ([raw file](https://github.com/cockroachdb/cockroach/blob/625193eee94b4fd02310b88ccea90aa121a085e5/docs/RFCS/leader_evaluated_raft.md#L35)):*
I think this a big deal for folks who are running large clusters where a change like this can save thousands of machines

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/6166)*
<!-- Sent from Reviewable.io -->
